contracts/Witch.sol

```diff
@@ -13,198 +13,358 @@ import "@yield-protocol/utils-v2/contracts/math/WDivUp.sol";
+    struct Line {
+        uint32 duration; // Time that auctions take to go to minimal price and stay there
+        uint64 proportion; // Proportion of the vault that is available each auction (1e18 = 100%)
+        uint64 initialOffer; // Proportion of collateral that is sold at auction start (1e18 = 100%)
     }
 
     struct Limits {
-        uint96 line;                                                    // Maximum concurrent auctioned collateral
-        uint24 dust;                                                    // Minimum collateral that must be left when buying, unless buying all
-        uint8 dec;                                                      // Multiplying factor (10**dec) for line and dust
-        uint128 sum;                                                    // Current concurrent auctioned collateral
+        uint96 max; // Maximum concurrent auctioned collateral
+        uint24 dust; // Minimum collateral that must be left when buying, unless buying all
+        uint8 dec; // Multiplying factor (10**dec) for max and dust
+        uint128 sum; // Current concurrent auctioned collateral
     }
 
-    ICauldron immutable public cauldron;
+    ICauldron public immutable cauldron;
     ILadle public ladle;
     mapping(bytes12 => Auction) public auctions;
-    mapping(bytes6 => Ilk) public ilks;
-    mapping(bytes6 => Limits) public limits;
+    mapping(bytes6 => mapping(bytes6 => Line)) public lines;
+    mapping(bytes6 => mapping(bytes6 => Limits)) public limits;
 
-    constructor (ICauldron cauldron_, ILadle ladle_) {
+    constructor(ICauldron cauldron_, ILadle ladle_) {
         cauldron = cauldron_;
         ladle = ladle_;
     }
 
-
     /// @dev Point to a different ladle
+    /// @param param Name of parameter to set (must be "ladle")
+    /// @param value Address of new ladle
     function point(bytes32 param, address value) external auth {
-        if (param == "ladle") ladle = ILadle(value);
-        else revert("Unrecognized parameter");
+        require(param == "ladle", "Unrecognized");
+        ladle = ILadle(value);
         emit Point(param, value);
     }
 
-    /// @dev Governance function to set:
+    /// @dev Governance function to set the parameters that govern how much collateral is sold over time.
+    /// @param ilkId Id of asset used for collateral
+    /// @param baseId Id of asset used for underlying
+    /// @param duration Time that auctions take to go to minimal price
+    /// @param proportion Vault proportion that is set for auction each time
+    /// @param initialOffer Proportion of collateral that is sold at auction start (1e18 = 100%)
+    function setLine(
+        bytes6 ilkId,
+        bytes6 baseId,
+        uint32 duration,
+        uint64 proportion,
+        uint64 initialOffer
+    ) external auth {
+        require(initialOffer <= 1e18, "InitialOffer above 100%");
+        require(proportion <= 1e18, "Proportion above 100%");
```

Comment:
I'd add `require(initialOffer >= 0.01e18)` or something like that to mitigate the chance of repeating the issue of config values with the wrong precision. Same for `proportion`

Reply:
Fixed, thanks. Allowed 0% as well for `initialProportion`